rpc/core/env.go

```diff
@@ -100,6 +107,69 @@ type Environment struct {
 	genChunks []string
 }
 
+// InitGenesisChunks checks whether it makes sense to create a cache of chunked
+// genesis data. It is called on Node startup and should be called only once.
+// Rules of chunking:
+// - if the genesis file's size is <= 16MB, then no chunking. An `Environment`
+// object will store a pointer to the genesis in its GenDoc field. Its genChunks
+// field will be set to nil. `/genesis` RPC API will return the GenesisDoc itself.
+// - if the genesis file's size is > 16MB, then use chunking. An `Environment`
+// object will store a slice of base64-encoded chunks in its genChunks field. Its
+// GenDoc field will be set to nil. `/genesis` RPC API will redirect users to use
+// the `/genesis_chunked` API.
+func (env *Environment) InitGenesisChunks() error {
+	if len(env.genChunks) > 0 {
+		// we already computed the chunks, return.
+		return nil
+	}
+
+	if env.GenDoc == nil {
+		// chunks not computed yet, but no genesis available.
+		// This should not happen.
+		return errors.New("pointer to genesis is nil")
+	}
+
+	data, err := cmtjson.Marshal(env.GenDoc)
+	if err != nil {
+		return err
+	}
+
+	// If genesis is less than 16MB, then no chunking.
+	// Keep a pointer to a GenesisDoc in env.GenDoc.
+	if len(data) <= genesisChunkSize {
+		env.genChunks = nil
+		return nil
+	}
+
+	var (
+		nChunks = (len(data) + genesisChunkSize - 1) / genesisChunkSize
+		chunks  = make([]string, nChunks)
+	)
+	for i := range nChunks {
+		var (
+			start = i * genesisChunkSize
+			end   = start + genesisChunkSize
+		)
+		if end > len(data) {
+			end = len(data)
+		}
+
+		// we make a copy here so that the original data isn't retained in memory.
+		// The GC will collect the data slice after exiting the function.
+		// Without the copy, it would keep it in memory.
+		chunk := make([]byte, end-start)
+		copy(chunk, data[start:end])
+		chunks[i] = base64.StdEncoding.EncodeToString(chunk)
+	}
+
+	env.genChunks = chunks
+
+	// we store the chunks; don't store a ptr to the genesis anymore.
+	env.GenDoc = nil
+
```

Comment:
If you apply my comment above than this has to be changed to 
```suggestion
	// we store the chunks; don't store a ptr to the genesis anymore.
	if len(chunks) > 1{
		env.GenDoc = nil
	}

```

Reply:
see my reply above 😄